include/openPMD/RecordComponent.hpp

```diff
@@ -282,7 +300,7 @@ template< typename T >
 inline void
 RecordComponent::storeChunk(std::shared_ptr<T> data, Offset o, Extent e)
 {
-    if( *m_isConstant )
+    if( *m_isConstant && !*m_isEmpty )
```

Comment:
shouldn't we throw a runtime error on `m_isEmpty` if `storeChunk` is called on it?

Reply:
My intention was to allow writing generic code that does not need to distinguish empty datasets from non-empty ones. The offset and extent are still sanity-checked, meaning that the call to `storeChunk` will throw anyway if a non-empty chunk is passed. If all checks pass, this operation is a no-op.

Reply:
So what do you consider a valid value for `data` in this case?

Reply:
I guess you are hinting at the following lines? 
```cpp
if( !data )
    throw std::runtime_error("Unallocated pointer passed during chunk store.");
```
You are raising a point that I didn't think about thoroughly enough apparently. C++ does not allow zero-sized arrays (I was under the impression it did), so the user will need to make some distinction. (Meaning I'll have to check my PIConGPU plugin)
I guess, we should let nullpointers slide in this case?

Reply:
I mean I follow your wish to have simple code paths, but just contextually it seams wrong to me to call `storeChunk` on a constant or empty record component.

If we decided to allow this, which I feel makes the API actually more corner-casy than simple, then the only logically allowed value for `data` should be a `nullptr` because everything else is quite surprising.

Reply:
Using `storeChunk` on a constant record component is still forbidden because its intention in the openPMD standard is to be an *alternative* to dealing with chunks.
Here, on the other hand, constant record components are merely an (arguably quick and dirty) *implementation* to allow storing chunks that are incidentally empty. 

But given this implementation, I see your point that `storeChunk` may give the wrong impression that the backend will eventually receive a dataset-related task which is definitely not the case. And since the C++ standard already makes a unified code path impossible, I would be ok with removing this functionality again if you prefer.

Related question: `storeChunk` does not yet support workflows where one rank contributes no data to a dataset that has data, right? We should consider this, if we decide to keep the change.

Reply:
(answered below since mail answer was wrapped around)